File: 2022/day07_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any, Literal

type FileType = Literal["F"] | Literal["D"]
type FileName = list[str]
type FileSize = int

type File = tuple[FileName, FileSize, FileType]

type State = tuple[FileName, list[File]]


def solution(content: list[str]) -> Any:
    files = parse(content)
    return sum(
        value
        for value in (size(files, file) for file in files if is_directory(file))
        if value < 100_000
    )


def is_directory(file: File) -> bool:
    _, _, tp = file
    return tp == "D"


def file_name(file: File) -> FileName:
    name, _, _ = file
    return name


def contains(f1: File, f2: File) -> bool:
    return "".join(file_name(f2)).endswith("".join(file_name(f1)))


def file_size(file: File) -> FileSize:
    _, fsize, _ = file
    return fsize


def size(files: list[File], file: File) -> FileSize:
    return sum(file_size(item) for item in files if contains(file, item))


def parse(content: list[str]) -> list[File]:
    state = ([], [])
    for line in content:
        state = parse_line(state, line)
    _, files = state
    return files


def parse_line(state: State, line: str) -> State:
    clean = line.strip()
    if clean.startswith("$ ls"):
        return state
    current, files = state
    if clean.startswith("$ cd .."):
        return (current[1:], files)
    if clean.startswith("$ cd "):
        return ([clean[5:], *current], files)
    if clean.startswith("dir "):
        return (current, [([clean[4:], *current], 0, "D"), *files])
    fsize, fname, *_ = clean.split(" ")
    return (current, [([fname, *current], int(fsize), "F"), *files])


def main() -> None:
    with open("day07.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()

File: 2022/day07-1.ts
type File = {
  name: string[];
  size: number;
  type: "F" | "D";
};

type State = {
  current: string[];
  files: File[];
};

function solution(content: string) {
  const files = parse(content.split("\n"));
  return files
    .filter((file) => file.type === "D")
    .map((item) => size(files, item))
    .filter((value) => value < 100_000)
    .reduce((acc, cur) => acc + cur, 0);
}

function contains(f1: File, f2: File): boolean {
  return f2.name.join("").endsWith(f1.name.join(""));
}

function size(files: File[], file: File): number {
  return files
    .filter((item) => contains(file, item))
    .reduce((acc, item) => item.size + acc, 0);
}

function parse(content: string[]): File[] {
  return content.reduce(parseLine, { current: [], files: [] }).files;
}

function parseLine(state: State, line: string): State {
  const clean = line.trim();
  if (clean.startsWith("$ ls")) return state;
  const { current, files } = state;
  if (clean.startsWith("$ cd ..")) return { current: current.slice(1), files };
  if (clean.startsWith("$ cd "))
    return { current: [clean.slice(5), ...current], files };
  if (clean.startsWith("dir "))
    return {
      current,
      files: [
        { name: [clean.slice(4), ...current], size: 0, type: "D" },
        ...files,
      ],
    };
  const [fsize, fname] = clean.split(" ");
  return {
    current,
    files: [
      { name: [fname, ...current], size: Number(fsize), type: "F" },
      ...files,
    ],
  };
}

Deno.readTextFile("day07.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
